### ISSUE TITLE
ICE involving catch_expr and never_type

### ISSUE THREAD

**Issue by crumblingstatue** [rust-lang/rust#43335]
```rust
#![feature(catch_expr, never_type)]

struct Error {
    cause: &'static str,
}

fn oops() -> Result<(), Error> {
    Err(Error { cause: "epic fail" } )
}

fn main() {
    let Err(Error { cause }): Result<!, Error> = do catch {
        oops()?;
        return; // OK
    };
    eprintln!("Nuclear meltdown initiated: {}", cause);
}
```

```
error: internal compiler error: broken MIR (_1 = ()): bad assignment (! = ()): Sorts(ExpectedFound { expected: !, found: () })
  --> src/main.rs:11:11
   |
11 |   fn main() {
   |  ___________^
12 | |     let Err(Error { cause }): Result<!, Error> = do catch {
13 | |         oops()?;
14 | |         return; // OK
15 | |     };
16 | |     eprintln!("Nuclear meltdown initiated: {}", cause);
17 | | }
   | |_^

```

**Comment by crumblingstatue** [comment]
Interestingly, if you add `return ();` at the end, the ICE disappears, leading to #43336.

**Comment by scottmcm** [comment]
Looks like this can be triggered with only `catch_expr` (without `never_type`) by using a `loop`:
```rust
fn anaconda<I: Iterator>(mut it: I) {
    let _: Result<(),()> = do catch {
        loop {
            it.next().ok_or(())?;
        }
    };
}
```
https://play.rust-lang.org/?gist=32c33a234dd7a8d6a8e961d4121b885e&version=nightly
```
error: internal compiler error: broken MIR in NodeId(4) (_3 = ()): bad assignment (! = ()): Sorts(ExpectedFound { expected: !, found: () })
```

**Comment by varkor** [comment]
Looks like this is fixed now (I suspect by https://github.com/rust-lang/rust/pull/47746).
